engine/src/main/java/org/terasology/rendering/dag/AbstractNode.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.rendering.dag;
+
+
+import com.google.api.client.util.Maps;
+import java.util.Map;
+import org.terasology.rendering.dag.states.StateType;
+import org.terasology.rendering.dag.states.StateValue;
+
+
+public abstract class AbstractNode implements Node {
+    protected Map<StateType, StateValue> desiredStates;
```

Comment:
Ok, first of all, as we delve into the details of this, it should be noted that OpenGL is a big giant state characterized by many attributes. Pretty much all opengl functions change the (big giant) OpenGL state by changing one or more of its attributes. So, we should stop thinking and communicating in terms of "states" and think in terms of "state changes", because there is really only one state and that changes through time.

Regarding this particular line I initially thought of a <string, object> map for this. But perhaps even that is too complicated. Perhaps what we need is a _set_ (or a list perhaps) of desired StateChange instances. 

Now, let's remind ourselves what a Node might want:

1. A node might require an opengl attribute set to its default. 
1.1. In the node, the desiredStateChanges set has no entry for this.
1.2. Before Node.process() is executed there is nothing to do because state management should always reset attributes to their default unless this node wanted something different
1.3 After Node.process() is executed there is nothing to do unless the next node wants the attribute set to a non-default value

2. A node might require an opengl attribute set to a non-default value.
2.1. In the node the desiredStateChanges set has an entry for this. This entry contains the non default value.
2.2. Before Node.process() is executed the state management system verifies if that particular attribute is already set to that value, if not the state change is added to the pipeline, to be executed before the process() method.
2.3. After Node.process is executed the state management system verifies if that particular attribute+value is needed by the next node. If it is needed, it doesn't do anything. If it isn't it adds a state change to the pipeline to revert the attribute to its default.

3. A node might be capable of handling any value of a particular opengl attribute.
3.1. In the node the desiredStateChanges set has an entry for this. This entry contains a null value.
3.2. Before Node.process is executed there is nothing to do because this node can handle whatever value the attribute is set to.
3.3 After Node.process is executed the attribute needs to be reset to its default unless the next node is also capable of handling any value OR it requires a different value, be it the default or otherwise.

In this context:

The RenderPipelineGenerator could add all the explicitly listed state changes to the pipeline, disregarding of them being redundant or not. For example it could generate:

```
set attribute "color" to (non default) "blue" - explicitely requested
process node A
set attribute "color" to (default) "red" - default behavior but redundant: next node needs "blue"
set attribute "color" to "blue". - explicitely requested but redundant: "blue" is the current value already
process Node B
set attribute "color" to (default) "red"
```

Then, it would be up to the RenderPipelineOptimizer to turn that pipeline into:

```
set attribute "color" to (non default) "blue"
process Node A
process Node B
set attribute "color" to (default) "red"
```

And this is the pipeline that would actually get executed. 

Notice I haven't said anything about what the pipelines actually look like. I imagine them as lists, but the first one, generated by the RenderPipelineGenerator could list StateChanges and Nodes directly as it is geared for analyses. The second one, generated by the RenderPipelineOptimizer might contain PipelineTask instances, geared instead for fast execution. I.e. a class implementing the interface PipelineTask might look like this:

```
class SetColorTask implements PipelineTask {
 
    private final Color color;

    // called by the RenderPipelineOptmizer, before it adds the task to the pipeline to execute.
    SetColorTask(StateChange colorStateChange) {
        color = colorStateChange.getValue();
    }

   // called by whatever is iterating the pipeline being executed
   public void execute() {
       GL99.setColorFunction(color);
   }
}
```

Sorry I have been very verbose, but I strongly believe this is what we should be aiming for.